expression/builtin_cast.go

```diff
@@ -1844,6 +1844,14 @@ func BuildCastFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldT
 // WrapWithCastAsInt wraps `expr` with `cast` if the return type of expr is not
 // type int, otherwise, returns `expr` directly.
 func WrapWithCastAsInt(ctx sessionctx.Context, expr Expression) Expression {
+	if expr.GetType().Tp == mysql.TypeEnum {
+		if col, ok := expr.(*Column); ok {
+			col = col.Clone().(*Column)
```

Comment:
Why also clone col here?

Reply:
In the expression `if(e>1, e,'a')`,  the first `e` evalType is Int, the second `e` evalType is String. If don't clone col, the two pointers will point to the same address..